File: npm_api/src/playbackToken.ts
/**
 * Local signing of viewer playback tokens for streams, VOD assets, and clips
 * with a JWT playback policy. The token is an ES256 JWS carrying the claims
 * the gateway's playback verifier checks: kid in the header (one of the
 * policy's allowed kids), exp (required), and optionally nbf, aud (matched
 * against requiredAudience), and custom claims (matched exactly against
 * requiredClaimsJson).
 */

export interface PlaybackTokenOptions {
  /** The PKCS#8 PEM private key createSigningKey returned. */
  privateKeyPem: string;
  /** The signing key's kid. */
  kid: string;
  /** Expiry as a Date or unix seconds. Set this or expiresIn. */
  expiresAt?: Date | number;
  /** Lifetime in seconds from issuedAt. */
  expiresIn?: number;
  /** Issue time as a Date or unix seconds; defaults to now. */
  issuedAt?: Date | number;
  /** Earliest use as a Date or unix seconds. */
  notBefore?: Date | number;
  /** Your viewer's identifier. */
  subject?: string;
  /** Audience; one value is sent as a string, several as an array. */
  audience?: string | ReadonlyArray<string>;
  /** Custom claims. They may not set exp, iat, nbf, aud, or sub. */
  claims?: Readonly<Record<string, unknown>>;
}

const registeredClaims = ["exp", "iat", "nbf", "aud", "sub"];

function unixSeconds(value: Date | number): number {
  return value instanceof Date ? Math.floor(value.getTime() / 1000) : Math.floor(value);
}

/** JSON with object keys sorted at every level, so every SDK signs the same bytes. */
export function canonicalJSON(value: unknown): string {
  if (Array.isArray(value)) {
    return `[${value.map(canonicalJSON).join(",")}]`;
  }
  if (value !== null && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    return `{${Object.keys(obj)
      .filter((k) => obj[k] !== undefined)
      .sort()
      .map((k) => `${JSON.stringify(k)}:${canonicalJSON(obj[k])}`)
      .join(",")}}`;
  }
  return JSON.stringify(value);
}

function base64url(bytes: Uint8Array): string {
  let binary = "";
  for (const b of bytes) {
    binary += String.fromCharCode(b);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function pemToDer(pem: string): Uint8Array {
  const m = /-----BEGIN PRIVATE KEY-----([\s\S]+?)-----END PRIVATE KEY-----/.exec(pem);
  if (!m?.[1]) {
    throw new TypeError(
      "privateKeyPem must be a PKCS#8 PEM (BEGIN PRIVATE KEY), as createSigningKey returns it"
    );
  }
  const binary = atob(m[1].replace(/\s+/g, ""));
  const der = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    der[i] = binary.charCodeAt(i);
  }
  return der;
}

/** The header and payload segments of a playback token, before signing. */
export function playbackTokenSigningInput(
  options: Omit<PlaybackTokenOptions, "privateKeyPem">
): string {
  if (!options.kid) {
    throw new TypeError("kid is required");
  }
  const issuedAt =
    options.issuedAt === undefined ? Math.floor(Date.now() / 1000) : unixSeconds(options.issuedAt);
  let exp: number;
  if (options.expiresAt !== undefined) {
    exp = unixSeconds(options.expiresAt);
  } else if (options.expiresIn !== undefined) {
    exp = issuedAt + Math.floor(options.expiresIn);
  } else {
    throw new TypeError(
      "expiresAt or expiresIn is required: the playback verifier rejects tokens without exp"
    );
  }
  const payload: Record<string, unknown> = {};
  for (const [name, value] of Object.entries(options.claims ?? {})) {
    if (registeredClaims.includes(name)) {
      throw new TypeError(`claims may not set ${name}; use the matching option`);
    }
    payload[name] = value;
  }
  payload.exp = exp;
  payload.iat = issuedAt;
  if (options.notBefore !== undefined) {
    payload.nbf = unixSeconds(options.notBefore);
  }
  if (options.subject !== undefined) {
    payload.sub = options.subject;
  }
  if (options.audience !== undefined) {
    const aud = typeof options.audience === "string" ? [options.audience] : [...options.audience];
    payload.aud = aud.length === 1 ? aud[0] : aud;
  }
  const encoder = new TextEncoder();
  const header = { alg: "ES256", kid: options.kid, typ: "JWT" };
  return `${base64url(encoder.encode(canonicalJSON(header)))}.${base64url(encoder.encode(canonicalJSON(payload)))}`;
}

/** Signs a viewer playback token with a tenant signing key (ES256). */
export async function signPlaybackToken(options: PlaybackTokenOptions): Promise<string> {
  const signingInput = playbackTokenSigningInput(options);
  const key = await crypto.subtle.importKey(
    "pkcs8",
    pemToDer(options.privateKeyPem) as BufferSource,
    { name: "ECDSA", namedCurve: "P-256" },
    false,
    ["sign"]
  );
  // WebCrypto returns the raw r||s signature JWS uses.
  const signature = await crypto.subtle.sign(
    { name: "ECDSA", hash: "SHA-256" },
    key,
    new TextEncoder().encode(signingInput) as BufferSource
  );
  return `${signingInput}.${base64url(new Uint8Array(signature))}`;
}

File: sdk_python/src/livepeer_frameworks/playback_token.py
"""Local signing of viewer playback tokens for streams, VOD assets, and clips
with a JWT playback policy. The token is an ES256 JWS carrying the claims the
gateway's playback verifier checks: kid in the header (one of the policy's
allowed kids), exp (required), and optionally nbf, aud (matched against
requiredAudience), and custom claims (matched exactly against
requiredClaimsJson)."""

from __future__ import annotations

import base64
import json
import time
from collections.abc import Mapping, Sequence
from datetime import datetime
from typing import Any

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import decode_dss_signature

_REGISTERED_CLAIMS = frozenset({"exp", "iat", "nbf", "aud", "sub"})


def _unix(value: datetime | int | float) -> int:
    if isinstance(value, datetime):
        return int(value.timestamp())
    return int(value)


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def canonical_json(value: Any) -> str:
    """JSON with object keys sorted at every level and no whitespace, so every SDK signs the same bytes."""
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def playback_token_signing_input(
    kid: str,
    *,
    expires_at: datetime | int | None = None,
    expires_in: int | None = None,
    issued_at: datetime | int | None = None,
    not_before: datetime | int | None = None,
    subject: str | None = None,
    audience: str | Sequence[str] | None = None,
    claims: Mapping[str, Any] | None = None,
) -> str:
    """The header and payload segments of a playback token, before signing."""
    if not kid:
        raise ValueError("kid is required")
    iat = int(time.time()) if issued_at is None else _unix(issued_at)
    if expires_at is not None:
        exp = _unix(expires_at)
    elif expires_in is not None:
        exp = iat + int(expires_in)
    else:
        raise ValueError("expires_at or expires_in is required: the playback verifier rejects tokens without exp")
    payload: dict[str, Any] = {}
    for name, value in (claims or {}).items():
        if name in _REGISTERED_CLAIMS:
            raise ValueError(f"claims may not set {name}; use the matching argument")
        payload[name] = value
    payload["exp"] = exp
    payload["iat"] = iat
    if not_before is not None:
        payload["nbf"] = _unix(not_before)
    if subject is not None:
        payload["sub"] = subject
    if audience is not None:
        aud = [audience] if isinstance(audience, str) else list(audience)
        payload["aud"] = aud[0] if len(aud) == 1 else aud
    header = {"alg": "ES256", "kid": kid, "typ": "JWT"}
    return f"{_b64url(canonical_json(header).encode())}.{_b64url(canonical_json(payload).encode())}"


def sign_playback_token(
    private_key_pem: str,
    kid: str,
    *,
    expires_at: datetime | int | None = None,
    expires_in: int | None = None,
    issued_at: datetime | int | None = None,
    not_before: datetime | int | None = None,
    subject: str | None = None,
    audience: str | Sequence[str] | None = None,
    claims: Mapping[str, Any] | None = None,
) -> str:
    """Signs a viewer playback token with a tenant signing key (the PKCS#8
    PEM createSigningKey returned) and its kid. Set expires_at or expires_in;
    issued_at defaults to now. claims may not set exp, iat, nbf, aud, or sub."""
    signing_input = playback_token_signing_input(
        kid,
        expires_at=expires_at,
        expires_in=expires_in,
        issued_at=issued_at,
        not_before=not_before,
        subject=subject,
        audience=audience,
        claims=claims,
    )
    key = serialization.load_pem_private_key(private_key_pem.encode(), password=None)
    if not isinstance(key, ec.EllipticCurvePrivateKey) or key.curve.name != "secp256r1":
        raise ValueError("private_key_pem must be a P-256 EC key, as createSigningKey returns it")
    der = key.sign(signing_input.encode("ascii"), ec.ECDSA(hashes.SHA256()))
    # JWS carries the raw r||s signature, not the DER encoding cryptography returns.
    r, s = decode_dss_signature(der)
    signature = r.to_bytes(32, "big") + s.to_bytes(32, "big")
    return f"{signing_input}.{_b64url(signature)}"
